keps/sig-multicluster/5339-clusterprofile-plugin-credentials/README.md

```diff
@@ -0,0 +1,619 @@
+
+1. Build the endpoint details of the cluster by reading properties of the ClusterProfile
+2. Call the CredentialsExternalProviders, following the same flow defined in [KEP 541](https://github.com/kubernetes/enhancements/blob/master/keps/sig-auth/541-external-credential-providers/README.md)
+3. Build the rest.Config and return it to the caller
+
+### Converting endpoint properties to cluster object
+
+The Cluster structure for the exec defined in KEP 541 assumes the following:
+```
+type Cluster struct {
+  // Server is the address of the kubernetes cluster (https://hostname:port).
+  Server string `json:"server"`
+  // TLSServerName is passed to the server for SNI and is used in the client to
+  // check server certificates against. If ServerName is empty, the hostname
+  // used to contact the server is used.
+  // +optional
+  TLSServerName string `json:"tls-server-name,omitempty"`
+  // InsecureSkipTLSVerify skips the validity check for the server's certificate.
+  // This will make your HTTPS connections insecure.
+  // +optional
+  InsecureSkipTLSVerify bool `json:"insecure-skip-tls-verify,omitempty"`
+  // CAData contains PEM-encoded certificate authority certificates.
+  // If empty, system roots should be used.
+  // +listType=atomic
+  // +optional
+  CertificateAuthorityData []byte `json:"certificate-authority-data,omitempty"`
+  // ProxyURL is the URL to the proxy to be used for all requests to this
+  // cluster.
+  // +optional
+  ProxyURL string `json:"proxy-url,omitempty"`
+  // Config holds additional config data that is specific to the exec
+  // plugin with regards to the cluster being authenticated to.
+  //
+  // This data is sourced from the clientcmd Cluster object's
+  // extensions[client.authentication.k8s.io/exec] field:
+  //
+  // clusters:
+  // - name: my-cluster
+  //   cluster:
+  //     ...
+  //     extensions:
+  //     - name: client.authentication.k8s.io/exec  # reserved extension name for per cluster exec config
+  //       extension:
+  //         audience: 06e3fbd18de8  # arbitrary config
+  //
+  // In some environments, the user config may be exactly the same across many clusters
+  // (i.e. call this exec plugin) minus some details that are specific to each cluster
+  // such as the audience.  This field allows the per cluster config to be directly
+  // specified with the cluster info.  Using this field to store secret data is not
+  // recommended as one of the prime benefits of exec plugins is that no secrets need
+  // to be stored directly in the kubeconfig.
+  // +optional
+  Config runtime.RawExtension `json:"config,omitempty"`
+}
+```
+
+This data must be created from the ClusterProfile, as described in the following section.
+
+### Standardizing the Endpoint Property
+
+In order to populate the Cluster object that the exec provider requires, we standardize the following properties.
+
+Property: `multicluster.sigs.io/execprovider-cluster`
```

Comment:
Why `execprovider` (which feels implementation-derived) rather than something aligned with the feature provided, such as `credprovider`?

Reply:
Ah, is the intent to match the `exec` section in `kubeconfig`?